File: hello-x86/lab1_intro/lab1_sample.c
#include <stdio.h>

void prologue() {
  printf("        .text\n");
  printf("        .globl _main\n");
  printf("_main:\n");
  printf("        pushq %%rbp\n");
  printf("        movq %%rsp, %%rbp\n");
  printf("#------------------------\n");
}

void epilogue() {
  printf("#------------------------\n");
  printf("        popq %%rsi      # pop the resultant value\n");
  printf("        leaq L1(%%rip), %%rdi\n");
  printf("        movq $0, %%rax\n");
  printf("        callq _printf  # print the resultant value\n");
  printf("        movq $0, %%rax\n");
  printf("        leaveq\n");
  printf("        retq\n");
  printf("        .data\n");
  printf("L1:     .string \"%%d\\n\"\n");
}

int codegen() {
  int n;
  char buf[32];
  prologue();

  fgets(buf, 32, stdin);
  sscanf(buf, "%d", &n);
  printf("        pushq $%d\n", n);

  fgets(buf, 32, stdin);
  sscanf(buf, "%d", &n);
  printf("        pushq $%d\n", n);
  printf("#------------------------\n");
  printf("        popq %%rbx\n");
  printf("        popq %%rax\n");
  printf("        addq %%rbx, %%rax\n");
  printf("        pushq %%rax\n");

  epilogue();
  return 0;
}

int main() {
  codegen();
  return 0;
}

File: hello-x86/lab1_intro/lab1_sample.ml
let prologue () =
  let code = "        .text\n"
           ^ "        .globl _main\n"
           ^ "_main:\n"
           ^ "        pushq %rbp\n"
           ^ "        movq %rsp, %rbp\n"
  in code

let epilogue () =
  let code = "        popq %rsi      # pop the resultant value\n"
           ^ "        leaq L1(%rip), %rdi\n"
           ^ "        movq $0, %rax\n"
           ^ "        callq _printf  # print the resultant value\n"
           ^ "        movq $0, %rax\n"
           ^ "        leaveq\n"
           ^ "        retq\n"
           ^ "        .data\n"
           ^ "L1:     .string \"%d\\n\"\n"
  in code


let codegen () =
  let str = read_line ()
  in let n = int_of_string str
  in let code1 = Printf.sprintf "        pushq $%d\n" n
  in let str = read_line ()
  in let n = int_of_string str
  in let code2 = Printf.sprintf "        pushq $%d\n" n
  in let code3 = "        popq %rbx\n"
               ^ "        popq %rax\n"
               ^ "        addq %rbx, %rax\n"
               ^ "        pushq %rax\n"
  in let code = prologue () ^ code1 ^ code2 ^ code3 ^ epilogue ()
  in Printf.printf "%s\n" code

let main () = codegen ()

let _ = main ()
